trace/trace_test.go

```diff
@@ -0,0 +1,110 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package trace
+
+import (
+	"testing"
+	"time"
+)
+
+func TestNew(t *testing.T) {
+	tests := map[string]*struct {
+		inputString   string
+		expectedTrace *Trace
+	}{
+		"Test 1: When string is empty": {
+			inputString: "",
+			expectedTrace: &Trace{
+				name:  "",
+				steps: nil,
+			},
+		},
+		"Test 2: When string is not empty": {
+			inputString: "test string",
+			expectedTrace: &Trace{
+				name:  "test string",
+				steps: nil,
+			},
+		},
+	}
+
+	for name, tt := range tests {
+		t.Run(name, func(t *testing.T) {
+			got := New(tt.inputString)
+			if got.name != tt.expectedTrace.name {
+				t.Fatalf("Failed %s:\n Expected %v \n Got %v \n", name, tt.expectedTrace, got)
+			}
+		})
+	}
+}
+
+func TestStep(t *testing.T) {
+	sampleTrace := &Trace{}
```

Comment:
I would create this variable within the test run `t.Run` otherwise it will be shared state across test cases.

Reply:
@andrewrynhard really nice catch !! So I did it for intentionally as the [`Step`](https://github.com/kubernetes/utils/blob/master/trace/trace.go#L47-L53) method actually appends the msg in the steps slice. So by keeping it global i will also able to check whether the length is also increasing when the method is called, as it is a negative test case. Trying to summarise, the result of `Step` method will contain a slice  of struct called `steps` and the last struct of the slice will contain the msg  with which it is called and the length of the steps will be increasing by 1 everytime the method is called so making these two things as base i tried to check whether the last struct contains the valid msg and length of the steps increases or not.

Reply:
In that case I think the test should take a slice of inputs and run Step on all of them per test case. Sharing state across test cases can make tests really difficult to debug.

Reply:
Okay will change the way as you requested thanks for the review

Reply:
addressed :+1: